src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -91,55 +81,120 @@ pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context
+    // 3. If the result is an odd number, the coefficient is negative. If the result is even, the coefficient is positive.
+    let side_length = 1 << num_bits;
+    let mut equality_checks = Vec::with_capacity(side_length);
+    for i in 0..side_length {
+        let mut check = S::ZERO;
+        for (j, combination) in precomputed_combinations.iter().enumerate() {
+            let bit: i8 = i8::from((i & j) == i);
+            if bit > 0 {
+                let sign = 1 - 2 * i8::try_from((!i & j).count_ones() & 1).unwrap();
+                if sign > 0 {
+                    check += combination;
+                } else {
+                    check -= combination;
+                }
+            }
+        }
+        equality_checks.push(check);
+    }
+    Ok(equality_checks)
+}
+
+//
+// Every equality check can be computed as a linear combination of coefficients.
+// For example, if we are given a 3-bit number `[x_3, x_2, x_1]`,
+// we can check if it is equal to 4, by computing:
+// $x_3(1-x_2)(1-x_1)$,
+// which expands to:
+// $x_3 - x_2*x_3 - x_1*x_3 + x_1*x_2*x_3$
+//
+// Since we need to check all possible values, it makes sense to pre-compute all
+// of the coefficients that are used across all of these equality checks. In this way,
+// we can minimize the total number of multiplications needed.
+//
+// We must pre-compute all combinations of bit values. The following loop does so.
+// It does so by starting with the array `[1]`.
+// The next step is to multiply this by `x_1` and append it to the end of the array.
+// Now the array is `[1, x_1]`.
+// The next step is to mulitply all of these values by `x_2` and append them to the end of the array.
+// Now the array is `[1, x_1, x_2, x_1*x_2]`
+// The next step is to mulitply all of these values of `x_3` and append them to the end of the array.
+// Now the array is `[1, x_1, x_2, x_1*x_2, x_3, x_1*x_3, x_2*x_3, x_1*x_2*x_3]`
+// This process continues for as many steps as there are bits of input.
+async fn pregenerate_all_combinations<F, C, S>(
+    ctx: C,
+    record_idx: usize,
+    input: &[Vec<S>],
+    num_bits: usize,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: SecretSharing<F>,
+{
+    let record_id = RecordId::from(record_idx);
+    let mut precomputed_combinations = Vec::with_capacity(1 << num_bits);
+    precomputed_combinations.push(ctx.share_of_one());
+    for (bit_idx, column) in input.iter().enumerate() {
+        let bit = &column[record_idx];
+        let step = 1 << bit_idx;
+        let num_children_to_add = precomputed_combinations.len();
+        precomputed_combinations.push(bit.clone());
+        if bit_idx == 0 {
+            continue;
+        }
+        let mut multiplication_results = try_join_all(
+            precomputed_combinations
+                .iter()
+                .enumerate()
+                .skip(1)
+                .take(num_children_to_add - 1)
```

Comment:
There might be a cleaner version of this.

You want to skip index 0, so that your step counter makes sense.  That seems right; you skip the bit where you multiply by one.  But you have to use `take()` to avoid multiplying bit by bit, which we don't need.  You could move the addition of the extra value until after the loop.

That means running the loop always, but I believe that `try_join_all(v)` for `v.is_empty()` is fast enough that the simplification might be worth it.

You can also reorder the enumerate and skip steps.  The only effect that has is on child_idx, which will be slightly smaller as a consequence.  But since `step = 1<<bit_idx` and `num_children_to_add < 1<<bit_idx`, you are never at risk of overflowing.